lib/HTML/Restrict.pm

```diff
@@ -151,16 +151,42 @@ sub _build_parser {
 
                     foreach my $source_type ( 'href', 'src', 'cite' ) {
 
-                        if ( $attr->{$source_type} ) {
-                            my $uri = URI->new( $attr->{$source_type} );
-                            if ( defined $uri->scheme ) {
+                        my $link = $attr->{$source_type};
+
+                        # Remove unprintable ASCII control characters, which
+                        # are 0..31. These characters are not valid in URLs,
+                        # but they can prevent the URI parser from recognizing
+                        # the scheme when they are used as leading characters.
+                        # Browsers will helpfully ignore some of them, meaning
+                        # that some of these characters (particularly 1..8 and
+                        # 14..31) can be used to defeat HTML::Restrict when
+                        # used as leading characters in a link.  In our case we
+                        # will strip them all regardless of where they are in
+                        # the URL. See
+                        # https://github.com/oalders/html-restrict/issues/30
+                        # https://url.spec.whatwg.org/
+                        # https://infra.spec.whatwg.org/#c0-control
+
+                        if ($link) {
+
+                            # C0 control chars (decimal 0..31)
+                            # sort of like $link =~ s/[[:^print:]]//g
+                            $link =~ s/[\00-\037]/ /g;
+
+                            my $url = URI->new($link)->as_string;
+
+                            ## The above regex doesn't strip the null byte
+                            $url =~ s{&#x?0+;}{}g;
```

Comment:
The issue here seems to be that `HTML::Parser` doesn't decode `&#0;` and friends. I think this is a better solution thank what you have:

```
diff --git a/lib/HTML/Restrict.pm b/lib/HTML/Restrict.pm
index 650dd8a..6040d7b 100644
--- a/lib/HTML/Restrict.pm
+++ b/lib/HTML/Restrict.pm
@@ -171,14 +171,9 @@ sub _build_parser {
 
                             # C0 control chars (decimal 0..31)
                             # sort of like $link =~ s/[[:^print:]]//g
-                            $link =~ s/[\00-\037]/ /g;
+                            $link =~ s/[\00-\037]|&#x?0+;/ /g;
 
-                            my $url = URI->new($link)->as_string;
-
-                            ## The above regex doesn't strip the null byte
-                            $url =~ s{&#x?0+;}{}g;
-
-                            $url = URI->new($url);
+                            my $url = URI->new($link);
                             if ( defined $url->scheme ) {
                                 delete $attr->{$source_type}
                                     if none { $_ eq $url->scheme }
```

Reply:
Yep, that's better.